File: Saltarelle.Test/Saltarelle.Test/Movies01.cs
using System;
using System.Runtime.CompilerServices;
using System.Collections.Generic;
using jQueryApi;

namespace Saltarelle.Test
{
	public class Movies01
	{
		List<Movie> movies ;
		int count;

		public static void Main()
		{
			jQuery.OnDocumentReady(new Movies01().Attach);
		}

		void Attach() {
			jQuery.OnDocumentReady(() => {
				
				PopulateMovies();			
				Templates templates = new Templates();
				Observable.Templates(templates);
				RenderMovieTemplate();
				LinkTextOnlyTemplate();
				LinkMovieTemplate();

				jQuery.Select("#buttonAddMovie").Click(AddMovie);
				jQuery.Select("#buttonRemoveLastMovie").Click(RemoveLastMovie);
				
			});
		}

		void AddMovie(jQueryEvent evt) {
			Observable.ToArrayObservable(movies)
				.Insert(movies.Count, 
				        new Movie{Name="new movie " + (++count).ToString(), ReleaseYear="YYYY"});
		}

		void RemoveLastMovie(jQueryEvent evt) {
			Observable.ToArrayObservable(movies).Remove(movies.Count-1,1);
		}

		void PopulateMovies(){

			count=0;
			movies = new List<Movie>();

			movies.Add(new Movie{ Name= "The Red Violin", ReleaseYear= "1998" });
			movies.Add(new Movie{ Name= "Eyes Wide Shut", ReleaseYear= "1999" });
			movies.Add(new Movie{ Name= "The Inheritance", ReleaseYear= "1976"});
		}


		void RenderMovieTemplate ( )
		{
			//"$.render.MovieTemplate(movies)"
			jQuery.Select("#renderedListContainer").Html(Observable.Render("MovieTemplate", movies));
		}


		void LinkTextOnlyTemplate ()
		{
			//$.link.TextOnlyTemplate(".linkedListPlaceholder" , movies, { target: "replace" });
			Observable.Link("TextOnlyTemplate",".linkedListPlaceholder", movies, new LinkOptions{Target="replace"});
		}

		//[InlineCode ("$.link.MovieTemplate({selector},{movies})")] // it works too
		void LinkMovieTemplate ()
		{
			Observable.Link("MovieTemplate","#linkedListContainer", movies);
		}


		[Serializable]
		[PreserveMemberCase]
		class Templates{

			public Templates(){
				MovieTemplate="#movieTemplate";
				TextOnlyTemplate="#textOnlyTemplate";
			}

			public string MovieTemplate {get;set;}
			public string TextOnlyTemplate {get;set;}
		}


		[Serializable]
		class Movie{
			public string Name {get;set;}
			public string ReleaseYear {get;set;}
		}

	}

}

File: Saltarelle.Test/MyApp.Web/js/Movies01.js
(function() {
	////////////////////////////////////////////////////////////////////////////////
	// Saltarelle.Test.Movies01.Movie
	var $Saltarelle_$Test_Movies01$Movie = function() {
	};
	$Saltarelle_$Test_Movies01$Movie.createInstance = function() {
		return $Saltarelle_$Test_Movies01$Movie.$ctor();
	};
	$Saltarelle_$Test_Movies01$Movie.$ctor = function() {
		var $this = {};
		$this.name = null;
		$this.releaseYear = null;
		return $this;
	};
	////////////////////////////////////////////////////////////////////////////////
	// Saltarelle.Test.Movies01.Templates
	var $Saltarelle_$Test_Movies01$Templates = function() {
	};
	$Saltarelle_$Test_Movies01$Templates.createInstance = function() {
		return $Saltarelle_$Test_Movies01$Templates.$ctor();
	};
	$Saltarelle_$Test_Movies01$Templates.$ctor = function() {
		var $this = {};
		$this.MovieTemplate = null;
		$this.TextOnlyTemplate = null;
		$this.MovieTemplate = '#movieTemplate';
		$this.TextOnlyTemplate = '#textOnlyTemplate';
		return $this;
	};
	////////////////////////////////////////////////////////////////////////////////
	// Saltarelle.Test.Movies01
	var $Saltarelle_Test_Movies01 = function() {
		this.$movies = null;
		this.$count = 0;
	};
	$Saltarelle_Test_Movies01.prototype = {
		$attach: function() {
			$(Function.mkdel(this, function() {
				this.$populateMovies();
				var templates = $Saltarelle_$Test_Movies01$Templates.$ctor();
				$.templates(templates);
				this.$renderMovieTemplate();
				this.$linkTextOnlyTemplate();
				this.$linkMovieTemplate();
				$('#buttonAddMovie').click(Function.mkdel(this, this.$addMovie));
				$('#buttonRemoveLastMovie').click(Function.mkdel(this, this.$removeLastMovie));
			}));
		},
		$addMovie: function(evt) {
			var $t2 = $.observable(this.$movies);
			var $t3 = this.$movies.length;
			var $t1 = $Saltarelle_$Test_Movies01$Movie.$ctor();
			$t1.name = 'new movie ' + (++this.$count).toString();
			$t1.releaseYear = 'YYYY';
			$t2.insert($t3, $t1);
		},
		$removeLastMovie: function(evt) {
			$.observable(this.$movies).remove(this.$movies.length - 1, 1);
		},
		$populateMovies: function() {
			this.$count = 0;
			this.$movies = [];
			var $t2 = this.$movies;
			var $t1 = $Saltarelle_$Test_Movies01$Movie.$ctor();
			$t1.name = 'The Red Violin';
			$t1.releaseYear = '1998';
			$t2.add($t1);
			var $t4 = this.$movies;
			var $t3 = $Saltarelle_$Test_Movies01$Movie.$ctor();
			$t3.name = 'Eyes Wide Shut';
			$t3.releaseYear = '1999';
			$t4.add($t3);
			var $t6 = this.$movies;
			var $t5 = $Saltarelle_$Test_Movies01$Movie.$ctor();
			$t5.name = 'The Inheritance';
			$t5.releaseYear = '1976';
			$t6.add($t5);
		},
		$renderMovieTemplate: function() {
			//"$.render.MovieTemplate(movies)"
			$('#renderedListContainer').html($.render.MovieTemplate(this.$movies));
		},
		$linkTextOnlyTemplate: function() {
			//$.link.TextOnlyTemplate(".linkedListPlaceholder" , movies, { target: "replace" });
			$.link.TextOnlyTemplate('.linkedListPlaceholder', this.$movies, { target: 'replace' });
		},
		$linkMovieTemplate: function() {
			$.link.MovieTemplate('#linkedListContainer', this.$movies);
		}
	};
	$Saltarelle_Test_Movies01.main = function() {
		var $t1 = new $Saltarelle_Test_Movies01();
		$(Function.mkdel($t1, $t1.$attach));
	};
	Type.registerClass(null, 'Saltarelle.$Test.Movies01$Movie', $Saltarelle_$Test_Movies01$Movie, Object);
	Type.registerClass(null, 'Saltarelle.$Test.Movies01$Templates', $Saltarelle_$Test_Movies01$Templates, Object);
	Type.registerClass(global, 'Saltarelle.Test.Movies01', $Saltarelle_Test_Movies01, Object);
	$Saltarelle_Test_Movies01.main();
})();
